### PR TITLE
Tt 97 grant or revoke roles to users

### PR DESCRIPTION
**Pull request by PaulRC-ioet**
Feature that allows users to grant and revoke roles

### CODE REVIEW

File: src/app/modules/users/store/user.reducers.ts [hunk 1]
```diff
@@ -33,6 +34,58 @@ export const userReducer = (state: UserState = initialState, action: UserActions
...
+        message: 'Revoke User Role Success',
+      };
+    }
+
+    case UserActionTypes.REVOKE_USER_ROLE_FAIL: {
+      return {
+        ...state,
+        data: state.data,
+        isLoading: false,
+        message: 'Something went wrong revoking user rol',
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.reducers.ts:88]: `role`*

File: src/app/modules/users/store/user.reducers.ts [hunk 2]
```diff
@@ -33,6 +34,58 @@ export const userReducer = (state: UserState = initialState, action: UserActions
...
+        message: 'Grant User Role Success',
+      };
+    }
+
+    case UserActionTypes.GRANT_USER_ROLE_FAIL: {
+      return {
+        ...state,
+        data: state.data,
+        isLoading: false,
+        message: 'Something went wrong granted user rol',
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.reducers.ts:62]: `granting user role`*

File: src/app/modules/users/store/user.reducer.spec.ts [hunk 3]
```diff
@@ -28,6 +29,70 @@ describe('userReducer', () => {
...
+    expect(state.data).toEqual([userGranted]);
+    expect(state.isLoading).toEqual(false);
+    expect(state.message).toEqual('Grant User Role Success');
+  });
+
+  it('on GrantRoleUserFail, should show a message with an error message', () => {
+    const action = new actions.GrantRoleUserFail('error');
+    const state = userReducer(initialState, action);
+
+    expect(state.message).toEqual('Something went wrong granted user rol');
```
Comment (Angeluz-07) [inline src/app/modules/users/store/user.reducer.spec.ts:60]: Update messages also here.

File: src/app/modules/users/components/users-list/users-list.component.ts [hunk 4]
```diff
@@ -18,19 +20,41 @@ export class UsersListComponent implements OnInit, OnDestroy, AfterViewInit {
...
+    this.store.dispatch(new LoadUsers());
     this.loadUsersSubscription = this.actionsSubject$
       .pipe(filter((action: any) => action.type === UserActionTypes.LOAD_USERS_SUCCESS))
       .subscribe((action) => {
         this.users = action.payload;
         this.rerenderDataTable();
       });
-    this.store.dispatch(new LoadUsers());
+
+    this.loadUsersSubscription = this.actionsSubject$
```
Comment (Angeluz-07) [inline src/app/modules/users/components/users-list/users-list.component.ts:46]: Not sure if this affects somehow but, It is ok to overwrite the subscription ?. Line 39 and 46.

**Review by Angeluz-07 — COMMENTED** [review]
LGTM. Only a few comments. I think we would need to check how to manage the code coverage comments.

File: src/app/modules/users/store/user.reducers.ts (hunk 2 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.reducers.ts]: Done.

File: src/app/modules/users/store/user.reducer.spec.ts (hunk 3 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/store/user.reducer.spec.ts]: Done

File: src/app/modules/users/components/users-list/users-list.component.ts (hunk 4 above)
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.ts]: I have changed to a new Subscription.

File: src/app/modules/users/components/users-list/users-list.component.html [hunk 5]
```diff
@@ -1,15 +1,37 @@
...
+  datatable
+  [dtOptions]="dtOptions"
+>
   <thead class="thead-blue">
     <tr class="d-flex">
-      <th class="col-6">User Email</th>
-      <th class="col-6">Names</th>
+      <th class="col">User Email</th>
+      <th class="col">Names</th>
+      <th class="col" *ngIf="isFlagOn">Roles</th>
```
[the remark this answers is not in the document]
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.html]: Its changed

File: src/app/modules/users/components/users-list/users-list.component.html [hunk 6]
```diff
@@ -1,15 +1,37 @@
...
     </tr>
   </thead>
   <app-loading-bar *ngIf="isLoading$ | async"></app-loading-bar>
   <tbody *ngIf="(isLoading$ | async) === false">
     <tr class="d-flex" *ngFor="let user of users">
-      <td class="col-sm-6">{{ user.email }}</td>
-      <td class="col-sm-6">{{ user.name }}</td>
+      <td class="col">{{ user.email }}</td>
+      <td class="col">{{ user.name }}</td>
+      <td class="col text-center" *ngIf="isFlagOn">
```
[the remark this answers is not in the document]
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.html]: Done

File: src/app/modules/users/components/users-list/users-list.component.ts [hunk 7]
```diff
@@ -18,19 +20,41 @@ export class UsersListComponent implements OnInit, OnDestroy, AfterViewInit {
   dtTrigger: Subject<any> = new Subject();
   @ViewChild(DataTableDirective, { static: false })
   dtElement: DataTableDirective;
+  dtOptions: any = {};
+  isFlagOn;
```
[the remark this answers is not in the document]
Comment (PaulRC-ioet) [reply src/app/modules/users/components/users-list/users-list.component.ts]: its changed

**Review by Angeluz-07 — APPROVED** [review]

**Review by josepato87 — APPROVED** [review]

**Pull request merged**